optax/contrib/_galore.py

```diff
@@ -0,0 +1,490 @@
+    Projection <https://arxiv.org/abs/2403.03507>`_, 2024
+  """
+  mu_dtype = utils.canonicalize_dtype(mu_dtype)
+
+  def init_fn(params: base.Params) -> GaLoreState:
+    # Use flattening to avoid brittle tuple-unpacking logic
+    if not isinstance(rank, int):
+      raise TypeError(f"`rank` must be an int, got {type(rank)}")
+    if rank <= 0:
+      raise ValueError(f"`rank` must be positive, got {rank}")
+
+    leaves, treedef = jax.tree.flatten(params)
+
+    m_list = []
+    v_list = []
+    proj_list = []
+
+    for p in leaves:
+      if p.ndim == 2:
+        m_dim, n_dim = p.shape
+        # Use left projection when m >= n, right when m < n
+        # This minimizes moment memory: left gives (r, n), right gives (m, r)
+        use_left = m_dim >= n_dim
+        effective_rank = min(rank, m_dim, n_dim)
+
+        if use_left:
+          projector_shape = (m_dim, effective_rank)
+          moment_shape = (effective_rank, n_dim)
+        else:
+          projector_shape = (n_dim, effective_rank)
+          moment_shape = (m_dim, effective_rank)
+
+        projector = jnp.zeros(projector_shape, dtype=p.dtype)
+        m = jnp.zeros(moment_shape, dtype=mu_dtype or p.dtype)
+        v = jnp.zeros(moment_shape, dtype=mu_dtype or p.dtype)
+      else:
+        # Non-2D parameters: use full-rank logic with dummy projector
+        projector = jnp.zeros((0,0), dtype=p.dtype)
+        m = jnp.zeros_like(p, dtype=mu_dtype or p.dtype)
+        v = jnp.zeros_like(p, dtype=mu_dtype or p.dtype)
+
+      m_list.append(m)
+      v_list.append(v)
+      proj_list.append(projector)
+
+    return GaLoreState(
+        count=jnp.zeros([], jnp.int32),
+        m=treedef.unflatten(m_list),
+        v=treedef.unflatten(v_list),
+        projector=treedef.unflatten(proj_list),
+    )
+
+  def update_fn(
+      updates: base.Updates,
+      state: GaLoreState,
+      params: Optional[base.Params] = None,
+  ) -> tuple[base.Updates, GaLoreState]:
+    """Apply GaLore update."""
+    del params
+    count = state.count
+    count_inc = numerics.safe_int32_increment(count)
+
+    # Check if we should update projection matrices
+    should_update_proj = (count % update_proj_gap) == 0
+
+    def update_2d_left(grad, m, v, projector):
+      """Update for 2D parameter with left projection."""
+      # Compute effective rank from gradient dimensions (not projector shape
+      # since projector could be (0,0) for non-2D params during tracing)
+      effective_rank = min(rank, grad.shape[0], grad.shape[1])
+      original_dtype = grad.dtype
+
+      # Update projector if needed
+      new_projector = jax.lax.cond(
+          should_update_proj,
+          lambda: _get_orthogonal_matrix_left(grad, effective_rank),
+          lambda: projector,
+      )
+
+      # Project gradient to low-rank subspace
+      low_rank_grad = _project_gradient_left(grad, new_projector)
+
+      # Adam update in low-rank space
```

Comment:
The repo for the paper [https://github.com/jiaweizzhao/GaLore](https://github.com/jiaweizzhao/GaLore) mentions that the core update can be any optimizer, making GaLore more of a composed optimizer.

Can you take a look at allowing this implementation to take any optax gradient transformation here instead of hard-coding Adam?